libafl/src/executors/hooks/intel_pt.rs

```diff
@@ -0,0 +1,92 @@
+use std::{
+    ptr::slice_from_raw_parts_mut,
+    string::{String, ToString},
+};
+
+use libafl_bolts::intel_pt::IntelPT;
+use libipt::{Asid, Image, SectionCache};
+use serde::Serialize;
+use typed_builder::TypedBuilder;
+
+use crate::{
+    executors::{hooks::ExecutorHook, HasObservers},
+    inputs::UsesInput,
+    Error,
+};
+
+/// Info of a binary's section that can be used during `Intel PT` traces decoding
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Section {
+    /// Path of the binary
+    pub file_path: String,
+    /// Offset of the section in the file
+    pub file_offset: u64,
+    /// Size of the section
+    pub size: u64,
+    /// Start virtual address of the section once loaded in memory
+    pub virtual_address: u64,
+}
+
+/// Hook to enable Intel Processor Trace (PT) tracing
+#[derive(Debug, TypedBuilder)]
+pub struct IntelPTHook {
+    #[builder(default = IntelPT::builder().build().unwrap())]
+    intel_pt: IntelPT,
+    #[builder(setter(transform = |sections: &[Section]| sections_to_image(sections).unwrap()))]
+    image: (Image<'static>, SectionCache<'static>),
+    map_ptr: *mut u8,
+    map_len: usize,
+}
+
+impl<S> ExecutorHook<S> for IntelPTHook
+where
+    S: UsesInput + Serialize,
+{
+    fn init<E: HasObservers>(&mut self, _state: &mut S) {}
+
+    fn pre_exec(&mut self, _state: &mut S, _input: &S::Input) {
+        self.intel_pt.enable_tracing().unwrap();
+    }
+
+    fn post_exec(&mut self, _state: &mut S, _input: &S::Input) {
+        self.intel_pt.disable_tracing().unwrap();
+
+        let slice = unsafe { &mut *slice_from_raw_parts_mut(self.map_ptr, self.map_len) };
+        let _ = self
+            .intel_pt
+            .decode_traces_into_map(&mut self.image.0, slice)
+            .inspect_err(|e| log::warn!("Intel PT trace decoding failed: {e}"));
+    }
+}
+
+// It would be nice to have this as a `From<IntoIter<Section>>`, but Rust's orphan rule doesn't like
```

Comment:
2 comments there:
- instead of `From<IntoIter<T>>` you can simply implement `FromIter<T>`, i think this is what you are looking for
- it's possible to implement it by wrapping with a `pub struct SectionWrapper(Section)` annotated with `#[repr(transparent)]`, it should work as intended.

Reply:
My comment is not precise here sorry, should be a `TryFrom` since returns a `Result`. And `TryFromIter`  doesn't exist